### PR TITLE
Chain configuration of ldProgram

### PR DESCRIPTION
**Pull request by erikd**
Cabal currently adds the `-r` (relocatable) flag to some invocations of `ldProgram`, but `lld` (commonly used on WIndows) does not understand the `-r` flag. This PR just detects if `ldProgram` accepts `-r` and conditionally adds it, only if it is supported.

Include the following checklist in your PR:

* [x] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#other-conventions).
* [ ] Any changes that could be relevant to users [have been recorded in the changelog](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#changelog).
* [ ] The documentation has been updated, if necessary.
* [ ] [Manual QA notes](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#qa-notes) have been included.

Bonus points for added automated tests!

### CODE REVIEW

**Review by andreabedini — APPROVED** [review]
Much gried was caused by `Distribution.Simple.GHC.Internal` overriding `programPostConf` set in `Distribution.Simple.Program.Builtin`.
I cannot shake the feeing this is an area that could be improved but this is ok for now IMHO.

**Comment by Mikolaj** [comment]
Is there an issue ticket connected to this PR? If not, could you add some motivation and explanation to the PR description? Thank you.

**Comment by andreabedini** [comment]
@erikd I understand this is a part of a series of patches towards a goal but @Mikolaj is right in pointing out it's not clear where this goal is explicitated :)

I found these links
https://github.com/haskell/cabal/pull/9236
https://github.com/haskell/cabal/pull/9229
I think something was mentioned in https://github.com/haskell/cabal/pull/9226
Then there's https://github.com/input-output-hk/devx-aux/issues/143

Could you write a summary of what you are aiming for?

**Comment by erikd** [comment]
@Mikolaj @andreabedini The main issue is that the `lld` linker (which I think is most commonly used on Windows) does not understand the `-r` flag. This PR simply detects if the linker supports the `-relocateable` flag and only adds `-r` when that flag is supported.

Does that clear things up?

**Comment by angerman** [comment]
This PR makes cabal agnostic to the linker. For GHCs using lld it will not try to create relocatabled pre-linked ghci libs. For GHCs using GNU ld (or any other linker that supports -r, potentially lld in the future) it permits the creation pre-linked ghci objects. These are significantly faster to load wirh our static rts linker in GHC.

While GHC linked to control the toolchain it ships with for native deployments. This PR makes cabal agnostic to the used toolchain, and especially prevents breakage for cross compiled GHCs when using the gun toolchain.

**Comment by erikd** [comment]
Rebased and force pushed.

**Review by geekosaur — APPROVED** [review]

**Comment by bgamari** [comment]
@erikd, can you please edit the MR description to provide an overview of what issue this MR addresses and how it does so?

**Comment by erikd** [comment]
@bgamari What is an MR? I assume you do not mean PR, because you said MR twice.

**Comment by Kleidukos** [comment]
@erikd MR is the term used by Gitlab

**Comment by bgamari** [comment]
Yes, apologies. MR stands for "merge request" and is GitLab's terminology for "pull request".

**Comment by erikd** [comment]
Updated.

File: Cabal/src/Distribution/Simple/Program/Ld.hs
Comment (bgamari) [inline Cabal/src/Distribution/Simple/Program/Ld.hs:111]: I'm a bit lost here. It seems to me that this will simply attempt to call `ld` without `-r` to join objects when `-r` is not supported. This is, AFAICT, not correct and will at best result in linking (undefined reference) errors, and at worst result in an invalid (non-relocatable) object.

Comment (erikd) [reply Cabal/src/Distribution/Simple/Program/Ld.hs:111]: If `ldProgram` doesn’t have `-r`, we shouldn’t even try to merge objects.

Comment (erikd) [reply Cabal/src/Distribution/Simple/Program/Ld.hs]: Updated to call `error` if `combineObjectFiles` is called with an `ldProgram` that does not support relocatable output.

Comment (bgamari) [reply Cabal/src/Distribution/Simple/Program/Ld.hs]: @Mikolaj is `error` the best way to fail here?

Comment (erikd) [reply Cabal/src/Distribution/Simple/Program/Ld.hs]: I had a bit of a look, but `error` is used quite extensively elsewhere in the `Cabal` library.

Comment (andreabedini) [reply Cabal/src/Distribution/Simple/Program/Ld.hs]: I believe `dieWithException` is the standard after #9219

File: Cabal/src/Distribution/Simple/Errors.hs
Comment (erikd) [inline Cabal/src/Distribution/Simple/Errors.hs:305]: Could not find any information about how this number is chosen. Found another relocation based exception and chose an unused number nearby.

Comment (bgamari) [reply Cabal/src/Distribution/Simple/Errors.hs:305]: @SuganyaAK, @sclv, indeed we should make sure that this gets documented somehow. Would one of you be willing to do so?

Comment (andreabedini) [reply Cabal/src/Distribution/Simple/Errors.hs:305]: I think the relevant discussion is https://github.com/haskell/cabal/issues/8543

**Comment by erikd** [comment]
@bgamari Ping? I think I need your approval before this can be merged.

**Review by angerman — APPROVED** [review]
🎉 no special casing, but feature detection!

File: Cabal/src/Distribution/Simple/Errors.hs
Comment (bgamari) [inline Cabal/src/Distribution/Simple/Errors.hs:798]: The wording of this error seems wrong. The linker definitely does support relocations (e.g. nonrelocatable output still has dynamic relocations). It is rather that it doesn't support producing relocatable object files.

**Review by bgamari — APPROVED** [review]

**Comment by andreabedini** [comment]
There's four approvals now, @erikd you can add the "merge-me" label when you are ready.

File: Cabal/src/Distribution/Simple/Program/Ld.hs
Comment (Kleidukos) [inline Cabal/src/Distribution/Simple/Program/Ld.hs:79]: Sorry for barging in at the eleventh hour, this is not a blocking request but would you mind annotating this comment with `NOTE`?

**Comment by erikd** [comment]
Removed the "mergeme" label because we may not actually be ready to merge.

**Comment by erikd** [comment]
Hope to replace this PR with https://github.com/haskell/cabal/pull/9443

**Comment by erikd** [comment]
This is redundant now that #9443 is merged.

**Pull request closed without merging**
